Documentation/Coherency.md

```diff
@@ -0,0 +1,74 @@
+- We are not necessarily shipping the exact bits that an official release would.
+- Servicing releases *should* be incoherent.  We do not ship non-serviced packages, so a valid official SDK could have some 3.0.1 bits and some 3.0.0 bits.
+	- This also means that we do not get all the packages we need from our forced-coherent build.
+- Very few official builds are coherent during development.  Builds diverge from coherence within days after release and soon become massively incoherent (dependency graph reports 44 distinct repo-shas for today's CoreFX build, and 181 distinct repo-shas for core-sdk).  We would like to support daily builds, which means that we have to deal with incoherence during the development process.
+
+# Solutions Considered #
+------------------------
+We have considered a few solutions:
+
+## Continue with traditional source-build coherence ##
+This would mean that we would continue to force coherence on every build, using our traditional method of passing around PackageVersions.props as the build progresses.
+Pros:
+- Simple.
+- Already have this mechanism.
+
+Cons:
+- We continue to not ship bits that match official builds, except on releases.
+- We have to deal with the problem of not getting all packages for servicing releases:
+	-  Issues https://github.com/dotnet/source-build/issues/210 and https://github.com/dotnet/source-build/issues/923 are our earlier discussions.
+	-  If we turn on BuildAllConfigurations and BuildAllPackages, we will ship bits that do not exist except on RHEL machines.  If someone shares a project but not the packages directory, restoring these packages that do not officially exist will fail.
+	-  If we don't turn on BuildAllConfigurations and BuildAllPackages, we will continue to have a large number of prebuilt packages.
+
+## Build multiple versions of each repo ##
+This would mean we build CLI with its expected version of CoreFX, and core-setup with its own expected version of CoreFX.
+Pros:
+- We match the official bits very closely.
+- We don't have to deal with servicing considerations.
+
+Cons:
+- This explodes rather quickly and would have extremely long build times.
+- We have to be careful about cross-contaminating package feeds during the builds.
+
+## Store and download packages in later builds ##
+During each build, we would save source-built packages off in an Azure account or NuGet feed.  During later builds, we would restore these packages and use them in builds later in the dependency tree.
+Pros:
+- Improved build times.
+	- Fringe benefit: incremental builds would probably finally work.
+- Still lets us eliminate prebuilt packages (all packages stored are source-built).
+- We match the official bits very closely.
+
+Cons:
+- Have to be very careful to not contaminate the storage with prebuilt packages.
+- Need internet access during the build.
```

Comment:
I think this is a limitation of the implementation, rather than the strategy. Can't we put it in a separate package (meaning RPM/Deb) to transfer to the N+1 build?

Reply:
I was thinking about the production build.  Split this out into two cons to make it clearer.